### PR TITLE
feat(presets): add enable/disable toggle and update semantics

### LINKED ISSUE #1851: Preset system: Add update semantics (update/restore methods)
## Summary
Port the safer mutation patterns from extensions to presets for better registry management.

## Context
Extensions have (extensions.py:232-278):
- `update()` method for safe partial updates
- `restore()` method for rollback scenarios
- Deep copy isolation (lines 303, 314)

Presets only have `add()` method.

## Requirements
- Add `update()` method to PresetRegistry for safe partial updates
- Add `restore()` method for rollback scenarios
- Use deep copy in `get()` and `list()` methods
- Prevent accidental mutations during concurrent operations

## Related
- Reverse parity: extensions → presets

### LINKED ISSUE #1852: Preset system: Add enable/disable toggle functionality
## Summary
Add ability to temporarily disable presets without removing them, similar to extensions.

## Context
Extensions have `enable` and `disable` commands that toggle functionality without full removal.

Presets lack this capability - the only option is full removal.

## Requirements
- Add `enabled` field to preset registry (may already exist)
- Add `preset enable` CLI command
- Add `preset disable` CLI command
- Skip disabled presets in resolution
- Show enabled/disabled status in `preset list`

## Related
- Reverse parity: extensions → presets

### PR DESCRIPTION
**Pull request by mbachorik**
## Summary

- Add `preset enable` and `preset disable` CLI commands for toggling presets without removal
- Add `restore()` method to PresetRegistry for rollback scenarios
- Update `get()` and `list()` to return deep copies (prevents accidental mutation)
- Update `list_by_priority()` to filter disabled presets by default
- Disabled presets are skipped during template resolution

## Addresses

- Closes #1851 - Preset system: Add update semantics (update/restore methods)
- Closes #1852 - Preset system: Add enable/disable toggle functionality

## Defensive Programming

Following feedback from previous PR, this implementation includes:
- Input validation in `restore()` for None/non-dict metadata
- Corrupted registry entry handling in enable/disable commands
- Deep copy protection to prevent state mutation
- Tests for all edge cases including corrupted state

## Test plan

- [x] All 386 tests pass
- [x] Ruff linter clean
- [x] New tests for `restore()`, deep copy, enable/disable CLI, corrupted state handling

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mbachorik** [comment]
Addressed both review comments:

1. **Error message accuracy**: Changed from "non-empty dict" to "dict" to match actual validation
2. **Deep copy**: Now using `copy.deepcopy(metadata)` instead of `dict(metadata)`

Also applied the same fixes to `ExtensionRegistry.restore()` for parity, and added tests for both.

Added `/defensive-check` command to catch these issues before PRs.

File: src/specify_cli/presets.py [hunk 1]
```diff
@@ -306,32 +306,61 @@ def update(self, pack_id: str, updates: dict):
...
+
+        Args:
+            pack_id: Preset ID
+            metadata: Complete preset metadata including installed_at
+
+        Raises:
+            ValueError: If metadata is None or not a dict
+        """
+        if metadata is None or not isinstance(metadata, dict):
+            raise ValueError(f"Cannot restore '{pack_id}': metadata must be a non-empty dict")
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - changed error message to "metadata must be a dict" to accurately reflect the validation.

File: src/specify_cli/presets.py [hunk 2]
```diff
@@ -306,32 +306,61 @@ def update(self, pack_id: str, updates: dict):
...
+        self.data["presets"][pack_id] = dict(metadata)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - now using `copy.deepcopy(metadata)` to prevent caller mutation. Also applied same fix to ExtensionRegistry for parity.

File: src/specify_cli/__init__.py [hunk 3]
```diff
@@ -2405,6 +2405,87 @@ def preset_set_priority(
...
+
+    if metadata.get("enabled", True):
+        console.print(f"[yellow]Preset '{pack_id}' is already enabled[/yellow]")
+        raise typer.Exit(0)
+
+    # Enable the preset
+    manager.registry.update(pack_id, {"enabled": True})
+
+    console.print(f"[green]✓[/green] Preset '{pack_id}' enabled")
+    console.print("\nTemplates from this preset will now be included in resolution.")
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/__init__.py:2445]: Added informational note to output: "Note: Previously registered commands/skills remain active."

Full command/skill unregistration on disable (with re-registration on enable) would be a larger change - tracking as a potential follow-up enhancement.

File: src/specify_cli/__init__.py [hunk 4]
```diff
@@ -2405,6 +2405,87 @@ def preset_set_priority(
...
+    if not metadata.get("enabled", True):
+        console.print(f"[yellow]Preset '{pack_id}' is already disabled[/yellow]")
+        raise typer.Exit(0)
+
+    # Disable the preset
+    manager.registry.update(pack_id, {"enabled": False})
+
+    console.print(f"[green]✓[/green] Preset '{pack_id}' disabled")
+    console.print("\nTemplates from this preset will be skipped during resolution.")
+    console.print(f"To re-enable: specify preset enable {pack_id}")
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/__init__.py:2488]: Added clarifying note to output: "Note: Previously registered commands/skills remain active until preset removal."

This makes it clear that disable only affects template resolution. Full artifact unregistration could be a follow-up enhancement.

File: src/specify_cli/extensions.py [hunk 5]
```diff
@@ -365,6 +373,9 @@ def list_by_priority(self) -> List[tuple]:
         for ext_id, meta in extensions.items():
             if not isinstance(meta, dict):
                 continue
+            # Skip disabled extensions unless explicitly requested
+            if not include_disabled and not meta.get("enabled", True):
+                continue
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py:426]: Fixed - `_get_all_extensions_by_priority()` now uses `include_disabled=True` to build `registered_extension_ids`, then explicitly skips disabled entries in the output. This prevents disabled extensions from being picked up via the unregistered-dir scan.

File: tests/test_extensions.py [hunk 6]
```diff
@@ -2500,6 +2542,40 @@ def test_list_by_priority_invalid_priority_defaults(self, temp_dir):
...
+        registry.add("ext-enabled", {"version": "1.0.0", "enabled": True, "priority": 5})
+        registry.add("ext-disabled", {"version": "1.0.0", "enabled": False, "priority": 1})
+
+        # Include disabled
+        by_priority = registry.list_by_priority(include_disabled=True)
+        ext_ids = [p[0] for p in by_priority]
+        assert "ext-enabled" in ext_ids
+        assert "ext-disabled" in ext_ids
+        # Disabled ext has lower priority number, so it comes first when included
+        assert ext_ids[0] == "ext-disabled"
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply tests/test_extensions.py:2611]: Added two integration tests in test_presets.py:
- `test_resolve_disabled_extension_templates_skipped` - verifies disabled extension templates return None
- `test_resolve_disabled_extension_not_picked_up_as_unregistered` - verifies disabled extensions are not picked up via unregistered dir fallback

File: src/specify_cli/presets.py [hunk 7]
```diff
@@ -306,32 +306,61 @@ def update(self, pack_id: str, updates: dict):
...
         Args:
             pack_id: Preset ID
 
         Returns:
-            Pack metadata or None if not found
+            Deep copy of preset metadata, or None if not found
         """
-        return self.data["presets"].get(pack_id)
+        entry = self.data["presets"].get(pack_id)
+        return copy.deepcopy(entry) if entry is not None else None
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - `get()` now returns `None` for corrupted (non-dict) entries in both `PresetRegistry` and `ExtensionRegistry`. Added `isinstance(entry, dict)` check before returning deep copy. Added tests for corrupted entries in both test files.

File: src/specify_cli/presets.py [hunk 8]
```diff
@@ -306,32 +306,64 @@ def update(self, pack_id: str, updates: dict):
...
 
+        Returns a deep copy of the presets mapping to prevent callers
+        from accidentally mutating nested internal registry state.
+
         Returns:
-            Dictionary of pack_id -> metadata
+            Dictionary of pack_id -> metadata (deep copies)
         """
-        return self.data["presets"]
+        return copy.deepcopy(self.data["presets"])
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - `list()` now validates that `presets`/`extensions` is a dict before returning. Returns empty dict for corrupted registry. Applied same fix to `ExtensionRegistry.list()` for parity. Added tests for both.

File: src/specify_cli/presets.py [hunk 9]
```diff
@@ -306,32 +306,67 @@ def update(self, pack_id: str, updates: dict):
...
         Returns:
-            Pack metadata or None if not found
+            Deep copy of preset metadata, or None if not found or corrupted
         """
-        return self.data["presets"].get(pack_id)
+        entry = self.data["presets"].get(pack_id)
+        # Return None for missing or corrupted (non-dict) entries
+        if entry is None or not isinstance(entry, dict):
+            return None
+        return copy.deepcopy(entry)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - now validates top-level structure before accessing. get() returns None if presets/extensions is missing or not a dict. restore() ensures the dict exists before writing.

File: src/specify_cli/presets.py [hunk 10]
```diff
@@ -306,32 +306,67 @@ def update(self, pack_id: str, updates: dict):
...
+        Args:
+            pack_id: Preset ID
+            metadata: Complete preset metadata including installed_at
+
+        Raises:
+            ValueError: If metadata is None or not a dict
+        """
+        if metadata is None or not isinstance(metadata, dict):
+            raise ValueError(f"Cannot restore '{pack_id}': metadata must be a dict")
+        self.data["presets"][pack_id] = copy.deepcopy(metadata)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py:342]: Fixed - now validates top-level structure before accessing. get() returns None if presets/extensions is missing or not a dict. restore() ensures the dict exists before writing.

File: src/specify_cli/extensions.py [hunk 11]
```diff
@@ -320,10 +325,13 @@ def get(self, extension_id: str) -> Optional[dict]:
...
         Returns:
-            Deep copy of extension metadata, or None if not found
+            Deep copy of extension metadata, or None if not found or corrupted
         """
         entry = self.data["extensions"].get(extension_id)
-        return copy.deepcopy(entry) if entry is not None else None
+        # Return None for missing or corrupted (non-dict) entries
+        if entry is None or not isinstance(entry, dict):
+            return None
+        return copy.deepcopy(entry)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py]: Fixed - now validates top-level structure before accessing. get() returns None if presets/extensions is missing or not a dict. restore() ensures the dict exists before writing.

File: src/specify_cli/presets.py [hunk 12]
```diff
@@ -306,32 +306,73 @@ def update(self, pack_id: str, updates: dict):
...
 
         Returns:
-            Pack metadata or None if not found
+            Deep copy of preset metadata, or None if not found or corrupted
         """
-        return self.data["presets"].get(pack_id)
+        packs = self.data.get("presets")
+        if not isinstance(packs, dict):
+            return None
+        entry = packs.get(pack_id)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py:360]: Fixed - _load() now validates json.load() returns a dict, is_installed() validates the presets/extensions key is a dict before membership check.

File: src/specify_cli/extensions.py [hunk 13]
```diff
@@ -320,10 +328,16 @@ def get(self, extension_id: str) -> Optional[dict]:
...
         Returns:
-            Deep copy of extension metadata, or None if not found
+            Deep copy of extension metadata, or None if not found or corrupted
         """
-        entry = self.data["extensions"].get(extension_id)
-        return copy.deepcopy(entry) if entry is not None else None
+        extensions = self.data.get("extensions")
+        if not isinstance(extensions, dict):
+            return None
+        entry = extensions.get(extension_id)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py:350]: Fixed - _load() now validates json.load() returns a dict, is_installed() validates the presets/extensions key is a dict before membership check.

File: src/specify_cli/__init__.py [hunk 14]
```diff
@@ -2405,6 +2405,89 @@ def preset_set_priority(
...
+    manager = PresetManager(project_root)
+
+    # Check if preset is installed
+    if not manager.registry.is_installed(pack_id):
+        console.print(f"[red]Error:[/red] Preset '{pack_id}' is not installed")
+        raise typer.Exit(1)
+
+    # Get current metadata
+    metadata = manager.registry.get(pack_id)
+    if metadata is None or not isinstance(metadata, dict):
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/__init__.py:2433]: Fixed - _load() now validates json.load() returns a dict, is_installed() validates the presets/extensions key is a dict before membership check.

File: src/specify_cli/__init__.py [hunk 15]
```diff
@@ -2405,6 +2405,89 @@ def preset_set_priority(
...
+    specify_dir = project_root / ".specify"
+    if not specify_dir.exists():
+        console.print("[red]Error:[/red] Not a spec-kit project (no .specify/ directory)")
+        console.print("Run this command from a spec-kit project root")
+        raise typer.Exit(1)
+
+    manager = PresetManager(project_root)
+
+    # Check if preset is installed
+    if not manager.registry.is_installed(pack_id):
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/__init__.py:2468]: Fixed - _load() now validates json.load() returns a dict, is_installed() validates the presets/extensions key is a dict before membership check.

File: src/specify_cli/presets.py [hunk 16]
```diff
@@ -238,7 +238,14 @@ def _load(self) -> dict:
...
         try:
             with open(self.registry_path, 'r') as f:
-                return json.load(f)
+                data = json.load(f)
+            # Validate loaded data is a dict (handles corrupted registry files)
+            if not isinstance(data, dict):
+                return {
+                    "schema_version": self.SCHEMA_VERSION,
+                    "presets": {}
+                }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py:247]: Fixed - _load() now normalizes presets/extensions field to {} if not a dict, making corrupted registries recoverable.

File: src/specify_cli/extensions.py [hunk 17]
```diff
@@ -222,7 +222,14 @@ def _load(self) -> dict:
...
         try:
             with open(self.registry_path, 'r') as f:
-                return json.load(f)
+                data = json.load(f)
+            # Validate loaded data is a dict (handles corrupted registry files)
+            if not isinstance(data, dict):
+                return {
+                    "schema_version": self.SCHEMA_VERSION,
+                    "extensions": {}
+                }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py:231]: Fixed - _load() now normalizes presets/extensions field to {} if not a dict, making corrupted registries recoverable.

File: src/specify_cli/presets.py [hunk 18]
```diff
@@ -1466,12 +1523,17 @@ def _get_all_extensions_by_priority(self) -> list[tuple[int, str, dict | None]]:
             return []
 
         registry = ExtensionRegistry(self.extensions_dir)
-        registered_extensions = registry.list_by_priority()
-        registered_extension_ids = {ext_id for ext_id, _ in registered_extensions}
+        # Get ALL registered extensions (including disabled) to know which dirs are tracked
+        all_registered = registry.list_by_priority(include_disabled=True)
+        registered_extension_ids = {ext_id for ext_id, _ in all_registered}
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - now using registry.list().keys() for tracking all extensions (including corrupted). Test assertion tightened.

File: tests/test_presets.py [hunk 19]
```diff
@@ -2001,3 +2205,189 @@ def test_mixed_legacy_and_new_presets_ordering(self, temp_dir):
...
+        assert result is not None
+        assert "test-pack" in str(result)
+
+        # Disable the preset
+        manager.registry.update("test-pack", {"enabled": False})
+
+        # Template should NOT be found when disabled
+        resolver2 = PresetResolver(project_dir)
+        result2 = resolver2.resolve("test-template", "template")
+        assert result2 is None or "test-pack" not in str(result2)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply tests/test_presets.py]: Fixed - now using registry.list().keys() for tracking all extensions (including corrupted). Test assertion tightened.

File: src/specify_cli/extensions.py [hunk 20]
```diff
@@ -320,10 +338,16 @@ def get(self, extension_id: str) -> Optional[dict]:
...
-        entry = self.data["extensions"].get(extension_id)
-        return copy.deepcopy(entry) if entry is not None else None
+        extensions = self.data.get("extensions")
+        if not isinstance(extensions, dict):
+            return None
+        entry = extensions.get(extension_id)
+        # Return None for missing or corrupted (non-dict) entries
+        if entry is None or not isinstance(entry, dict):
+            return None
+        return copy.deepcopy(entry)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py:354]: Fixed - added defensive check in ExtensionManager.remove() to handle None metadata. Now matches PresetManager.remove() pattern.

File: src/specify_cli/presets.py [hunk 21]
```diff
@@ -1466,12 +1523,20 @@ def _get_all_extensions_by_priority(self) -> list[tuple[int, str, dict | None]]:
             return []
 
         registry = ExtensionRegistry(self.extensions_dir)
-        registered_extensions = registry.list_by_priority()
-        registered_extension_ids = {ext_id for ext_id, _ in registered_extensions}
+        # Use raw registry keys to track ALL extensions (including corrupted entries)
+        # This prevents corrupted entries from being picked up as "unregistered" dirs
+        registered_extension_ids = set(registry.list().keys())
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - added lightweight keys() method, list() now filters corrupted entries, comment updated.

File: src/specify_cli/presets.py [hunk 22]
```diff
@@ -1466,12 +1523,20 @@ def _get_all_extensions_by_priority(self) -> list[tuple[int, str, dict | None]]:
...
+
+        # Get enabled extensions for resolution (list_by_priority skips corrupted/disabled)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - added lightweight keys() method, list() now filters corrupted entries, comment updated.

File: src/specify_cli/presets.py [hunk 23]
```diff
@@ -306,32 +316,73 @@ def update(self, pack_id: str, updates: dict):
...
+
         Returns:
-            Dictionary of pack_id -> metadata
+            Dictionary of pack_id -> metadata (deep copies), empty dict if corrupted
         """
-        return self.data["presets"]
+        packs = self.data.get("presets", {}) or {}
+        if not isinstance(packs, dict):
+            return {}
+        return copy.deepcopy(packs)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py]: Fixed - added lightweight keys() method, list() now filters corrupted entries, comment updated.

File: src/specify_cli/extensions.py [hunk 24]
```diff
@@ -332,9 +356,12 @@ def list(self) -> Dict[str, dict]:
...
 
         Returns:
-            Dictionary of extension_id -> metadata (deep copies)
+            Dictionary of extension_id -> metadata (deep copies), empty dict if corrupted
         """
-        return copy.deepcopy(self.data["extensions"])
+        extensions = self.data.get("extensions", {}) or {}
+        if not isinstance(extensions, dict):
+            return {}
+        return copy.deepcopy(extensions)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py]: Fixed - added lightweight keys() method, list() now filters corrupted entries, comment updated.

File: src/specify_cli/presets.py [hunk 25]
```diff
@@ -303,35 +317,95 @@ def update(self, pack_id: str, updates: dict):
         else:
             # If not present in existing, explicitly remove from merged if caller provided it
             merged.pop("installed_at", None)
-        self.data["presets"][pack_id] = merged
+        packs[pack_id] = merged
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/presets.py:320]: Fixed - `update()` now uses `copy.deepcopy(updates)` to prevent caller mutation of nested objects.

File: src/specify_cli/extensions.py [hunk 26]
```diff
@@ -283,7 +294,7 @@ def update(self, extension_id: str, metadata: dict):
         else:
             # If not present in existing, explicitly remove from merged if caller provided it
             merged.pop("installed_at", None)
-        self.data["extensions"][extension_id] = merged
+        extensions[extension_id] = merged
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply src/specify_cli/extensions.py:297]: Fixed - `update()` now uses `copy.deepcopy(metadata)` to prevent caller mutation of nested objects.

**Comment by mbachorik** [comment]
Ready for review.

**Comment by mnriem** [comment]
Thank you!
